File: easy/most_common_safe_word.go
package easy

import (
	"maps"
	"slices"
	"strings"
	"unicode"
)

type Concordance struct {
	Word      string
	Frequency int
}

func MostCommonSafeWord(paragraph string, banned []string) string {
	freq := make(map[string]Concordance)
	words := strings.FieldsFunc(strings.ToLower(paragraph), func(r rune) bool {
		return !unicode.IsLetter(r)
	})
	for _, word := range words {
		if !slices.Contains(banned, word) && word != "" {
			count := freq[word]
			count.Word = word
			count.Frequency++
			freq[word] = count
		}
	}
	result := slices.Collect(maps.Values(freq))

	slices.SortFunc(result, func(a, b Concordance) int {
		return b.Frequency - a.Frequency
	})
	if len(result) == 0 {
		return ""
	}
	return result[0].Word

}

File: easy/most_common_safe_word.py
import re


def most_common_safe_word(paragraph: str, banned: list[str]) -> str:
    freq = {}
    words = re.split(r"[!?'.,;\s]+", paragraph.lower())
    for word in words:
        if word not in banned and len(word) > 0:
            freq[word] = freq.get(word, 0) + 1
    return sorted(freq.items(), reverse=True, key=lambda item: item[1])[0][0]
